fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java

```diff
@@ -16,52 +16,132 @@
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.WorldChunk;
 
 @Mixin(ChunkRendererRegion.class)
 public abstract class MixinChunkRendererRegion implements RenderAttachedBlockView {
-    private HashMap<BlockPos, Object> fabric_renderDataObjects;
-
-    @Inject(at = @At("RETURN"), method = "<init>")
-    public void init(World world, int cxOff, int czOff, WorldChunk[][] chunks, BlockPos posFrom, BlockPos posTo, CallbackInfo info) {
-        HashMap<BlockPos, Object> map = new HashMap<>();
-
-        for (WorldChunk[] chunkA : chunks) {
-            for (WorldChunk chunkB : chunkA) {
-                for (Map.Entry<BlockPos, BlockEntity> entry: chunkB.getBlockEntities().entrySet()) {
-                    BlockPos entPos = entry.getKey();
-                    if (entPos.getX() >= posFrom.getX() && entPos.getX() <= posTo.getX()
-                            && entPos.getY() >= posFrom.getY() && entPos.getY() <= posTo.getY()
-                            && entPos.getZ() >= posFrom.getZ() && entPos.getZ() <= posTo.getZ()) {
-
-                        Object o = ((RenderAttachmentBlockEntity) entry.getValue()).getRenderAttachmentData();
-                        if (o != null) {
-                            map.put(entPos, o);
-                        }
-                    }
-                }
-            }
-        }
-
-        this.fabric_renderDataObjects = map;
-    }
-
-    @Override
-    public Object getBlockEntityRenderAttachment(BlockPos pos) {
-        return fabric_renderDataObjects.get(pos);
-    }
+	private Int2ObjectOpenHashMap<Object> fabric_renderDataObjects;
+
+	@Shadow
+	protected abstract int getIndex(BlockPos pos);
+
+	@Shadow
+	protected abstract int getIndex(int x, int y, int z);
+
+	@Inject(at = @At("RETURN"), method = "<init>")
+	public void init(World world, int cxOff, int czOff, WorldChunk[][] chunks, BlockPos posFrom, BlockPos posTo, CallbackInfo info) {
+		// instantiated lazily - avoids allocation for chunks without any data objects - which is most of them!
+		Int2ObjectOpenHashMap<Object> map = null;
+
+		for (WorldChunk[] chunkOuter : chunks) {
+			for (WorldChunk chunk : chunkOuter) {
+				// Hash maps in chunks should generally not be modified outside of client thread
+				// but does happen in practice, due to mods or inconsistent vanilla behaviors, causing 
+				// CMEs when we iterate the map.  (Vanilla does not iterate these maps when it builds
+				// the chunk cache and does not suffer from this problem.)
+				// 
+				// We handle this by trying to iterating the block entity map first, because that 
+				// usually works and is most efficient.  If we encounter a CME, we retry, querying the 
+				// map by iterating block positions. This method is a little slower but will not cause a CME.
+				try {
+					map = mapChunkFast(chunk, posFrom, posTo, map);
+				} catch (ConcurrentModificationException e) {
+					map = mapChunkSafely(chunk, posFrom, posTo, map);
```

Comment:
1. consider logging this cme for future debug purposes
2. clear the map before the safe mapping

Reply:
Not understanding what you mean be remove the map argument. it is needed in all cases for this to function properly.

Reply:
no... you always pass in a null for the map. just remove that param

Reply:
map is only guaranteed to be null on the first iteration - note that it is inside a loop